src/channel.cr

```diff
@@ -204,69 +260,86 @@ class Channel(T)
   end
 
   def receive_impl
-    @lock.sync do
-      receive_internal do
-        yield if @closed
-
-        value = uninitialized T
-        state = DeliveryState::None
-        @receivers << Receiver(T).new(Fiber.current, pointerof(value), pointerof(state), select_context: nil)
-        @lock.unsync do
-          Crystal::Scheduler.reschedule
-        end
-
-        case state
-        when DeliveryState::Delivered
-          value
-        when DeliveryState::Closed
-          yield
-        else
-          raise "BUG: Fiber was awaken without channel delivery state set"
-        end
+    receiver = Receiver(T).new
+
+    @lock.lock
+
+    state, value = receive_internal
+
+    case state
+    when DeliveryState::Delivered
+      @lock.unlock
+      value.as(T)
```

Comment:
this cast is needed because the distinguished value `UseDefault` is removed and not used in `receive_internal` when returning `{DeliveryState::Closed, nil}`. I like to be stricter here I think.

Reply:
Ok, I will make `UseDefault` back. Returning {DeliveryState::Closed, UseDefault} sounds better. Thanks.